pool/pool.go

```diff
@@ -248,44 +256,107 @@ func (x *wrapperPrm) setResponseInfoCallback(f func(sdkClient.ResponseMetaInfo)
 	x.responseInfoCallback = f
 }
 
-// setDialContext specifies context for client dial.
-func (x *wrapperPrm) setDialContext(ctx context.Context) {
-	x.dialCtx = ctx
-}
-
 // newWrapper creates a clientWrapper that implements the client interface.
-func newWrapper(prm wrapperPrm) (*clientWrapper, error) {
+func newWrapper(prm wrapperPrm) *clientWrapper {
+	var cl sdkClient.Client
 	var prmInit sdkClient.PrmInit
 	prmInit.SetDefaultPrivateKey(prm.key)
 	prmInit.SetResponseInfoCallback(prm.responseInfoCallback)
 
+	cl.Init(prmInit)
+
 	res := &clientWrapper{
-		key:                 prm.key,
+		client:              &cl,
 		clientStatusMonitor: newClientStatusMonitor(prm.address, prm.errorThreshold),
+		prm:                 prm,
 	}
 
-	res.client.Init(prmInit)
+	return res
+}
+
+// dial establishes a connection to the server from the NeoFS network.
+// Returns an error describing failure reason. If failed, the client
+// SHOULD NOT be used.
+func (c *clientWrapper) dial(ctx context.Context) error {
+	cl, err := c.getClient()
+	if err != nil {
+		return err
+	}
 
 	var prmDial sdkClient.PrmDial
-	prmDial.SetServerURI(prm.address)
-	prmDial.SetTimeout(prm.timeout)
-	prmDial.SetContext(prm.dialCtx)
+	prmDial.SetServerURI(c.prm.address)
+	prmDial.SetTimeout(c.prm.timeout)
+	prmDial.SetContext(ctx)
 
-	err := res.client.Dial(prmDial)
-	if err != nil {
-		return nil, fmt.Errorf("client dial: %w", err)
+	if err = cl.Dial(prmDial); err != nil {
+		c.setUnhealthy()
+		return err
 	}
 
-	return res, nil
+	return nil
+}
+
+// restartIfUnhealthy checks healthy status of client and recreate it if status is unhealthy.
+// Return current healthy status and indicating if status was changed by this function call.
+func (c *clientWrapper) restartIfUnhealthy(ctx context.Context) (healthy, changed bool) {
+	var wasHealthy bool
+	if _, err := c.endpointInfo(ctx, prmEndpointInfo{}); err == nil {
+		return true, false
+	} else if !errors.Is(err, ErrPoolClientUnhealthy) {
+		wasHealthy = true
+	}
+
+	var cl sdkClient.Client
+	var prmInit sdkClient.PrmInit
+	prmInit.SetDefaultPrivateKey(c.prm.key)
+	prmInit.SetResponseInfoCallback(c.prm.responseInfoCallback)
+
+	cl.Init(prmInit)
+
+	var prmDial sdkClient.PrmDial
+	prmDial.SetServerURI(c.prm.address)
+	prmDial.SetTimeout(c.prm.timeout)
+	prmDial.SetContext(ctx)
+
+	if err := cl.Dial(prmDial); err != nil {
+		c.setUnhealthy()
+		return false, wasHealthy
+	}
+
+	c.clientMutex.Lock()
+	c.client = &cl
+	c.clientMutex.Unlock()
+
+	if _, err := cl.EndpointInfo(ctx, sdkClient.PrmEndpointInfo{}); err != nil {
+		c.setUnhealthy()
+		return false, wasHealthy
+	}
+
+	c.setHealthy()
+	return true, !wasHealthy
+}
+
+func (c *clientWrapper) getClient() (*sdkClient.Client, error) {
```

Comment:
`getClient` is introduced to solve potential races with `c.client` access?

Reply:
Yes. We recreate client sometimes